src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -348,13 +432,62 @@ public void DoDisconnectIfNeeded()
                 }
 
                 Self._socket.Dispose();
+                Self.ReturnAllBlocks();
                 Self._log.ConnectionStop(Self._connectionId);
                 Complete();
             }
 
             public void Complete()
             {
-                Self.OnWriteCompleted(Buffers, WriteStatus, WriteError);
+                Self.OnWriteCompleted(_byteCount, WriteStatus, WriteError);
+            }
+
+            private void ReturnFullyWrittenBlocks()
+            {
+                var block = _lockedStart.Block;
+                while (block != _lockedEnd.Block)
+                {
+                    var returnBlock = block;
+                    block = block.Next;
+
+                    returnBlock.Unpin();
+                    returnBlock.Pool?.Return(returnBlock);
+                }
+            }
+
+            private void LockWrite()
```

Comment:
Only the writer can advance the _tail. Only the libuv thread can advance _head.

The libuv thread writes all the data between _head and a snapshot it takes of _tail. Once the write call is made _head is advanced to the _tail snapshot (_lockedEnd). Once the write is complete, all the blocks between the original _head (_lockedStart) and the _tailSnapshot (excluding the _tail block which might still be written to) are returned to the pool.

The socket is closed from the libuv thread. At this time it will take the _returnLock and set _head and _tail to _null after returning all blocks between _head and _tail inclusive.

If the writer is currently between a call to ProducingStart and ProducingComplete when the socket is closed, the _tail will not be immediately returned to the pool. Instead, what was the _tail will be stored in _returnFromOnProducingComplete. When ProducingComplete is then called, all the blocks starting _returnFromOnProducingComplete will be immediately returned to the pool.